docs: Add glossary to documentation (#1127)
aneoconsulting/ArmoniK: .docs/content/0.armonik/1.glossary.md

# Glossary

> **Client:** User-developed software that communicates with the ArmoniK Control Plane to submit a list of tasks to be executed (by one or several workers). It also retrieves results and any error.

Reviewer: ```suggestion
**Client:** User-developed software that communicates with the ArmoniK Control Plane to submit a list of tasks to be executed (by one or several workers) and retrieve results and errors.
```

> **Compute Plane:** Expression designating the set of compute pods, i.e. the pods running a Polling/Scheduling Agent + Worker pair within the Kubernetes cluster.

Reviewer 2: , i.e., 
Commas before and after

> **Container:** A container is a standard unit of software that packages up code and all its dependencies so the application runs quickly and reliably from one computing environment to another. (see Docker (documentation)[https://www.docker.com/resources/what-container/])

Reviewer: ```suggestion
**Container:** A container is a standard unit of software that packages up code and all its dependencies so the application runs quickly and reliably from one computing environment to another. (see Docker [documentation](https://www.docker.com/resources/what-container/))
```

> **Control Plane:** Expression designating a set of applications (distributed in different pods) that maintain the Kubernetes cluster as well as communications between certain components (queue, Redis, MongoDB) and the *client*.

Reviewer: ```suggestion
**Control Plane:** Set of applications (distributed in different pods) that maintain the Kubernetes cluster as well as communications between certain components (queue, Redis, MongoDB) and the *client*.
```

> **Control Plane:** Expression designating a set of applications (distributed in different pods) that maintain the Kubernetes cluster as well as communications between certain components (queue, Redis, MongoDB) and the *client*.

Reviewer 3: Should be clarified a little bit. There are 2 control planes: ArmoniK's and Kubernetes's

**Data dependency:** Input data for a given task that depends on another unique task. Data dependencies formalize dependencies between tasks.

**Data Plane:** Expression designating the set of software components running the various storage and database systems within ArmoniK.

> **Kubernetes:** Kubernetes is an open source container orchestration engine for automating deployment, scaling, and management of containerized applications. The open source project is hosted by the Cloud Native Computing Foundation ((CNCF)[https://www.cncf.io/about]) (see Kubernetes (documentation)[https://kubernetes.io/docs/home/#:~:text=Kubernetes%20is%20an%20open%20source,and%20management%20of%20containerized%20applications.]).

Reviewer: ```suggestion
**Kubernetes:** Kubernetes is an open source container orchestration engine for automating deployment, scaling, and management of containerized applications. The open source project is hosted by the Cloud Native Computing Foundation ([CNCF](https://www.cncf.io/about)) (see Kubernetes [documentation](https://kubernetes.io/docs/home/#:~:text=Kubernetes%20is%20an%20open%20source,and%20management%20of%20containerized%20applications.)).
```

Reviewer 4: no need for a comma before "and management"

> **MongoDB:** MongoDB is a document database designed for ease of application development and scaling (see MongoDB (documentation)[https://www.mongodb.com/docs/manual/]).

Reviewer: ```suggestion
**MongoDB:** MongoDB is a document database designed for ease of application development and scaling (see MongoDB [documentation](https://www.mongodb.com/docs/manual/)).
```

> **Node/Nœud:** Kubernetes runs a workload by placing containers into Pods to run on Nodes. A node may be a virtual or physical machine, depending on the cluster. Each node is managed by the Kubernetes control plane and contains the services necessary to run Pods. (see Kubernetes (documentation)[https://kubernetes.io/docs/concepts/architecture/nodes/]). 

Reviewer: ```suggestion
**Node:** Kubernetes runs a workload by placing containers into Pods to run on Nodes. A node may be a virtual or physical machine, depending on the cluster. Each node is managed by the Kubernetes control plane and contains the services necessary to run Pods. (see Kubernetes [documentation}(https://kubernetes.io/docs/concepts/architecture/nodes/)).
```

Reviewer 4: I don't see why the definition starts with Kubernetes. Start with " an node may be..." and add the first sentence at the end of the description.

> **Partition:** Logical segmentation of the Kubernetes cluster's pool of machines to distribute workloads according to usage.

Reviewer 3: I would precise that it's an ArmoniK feature. Like that, it can imply a Kubernetes feature

**Payload:** Input data for a task that does not depend on any other task.

**Pod:** Pods are the smallest deployable units of computing that one can create and manage in Kubernetes. A Pod is a group of one or more containers, with shared storage and network resources, and a specification for how to run the containers. A Pod's contents are always co-located and co-scheduled, and run in a shared context. A Pod models an application-specific "logical host": it contains one or more application containers which are relatively tightly coupled (see Kubernetes (documentation)[https://kubernetes.io/docs/concepts/workloads/pods/]).

> **Polling/Scheduling agent:** Containerized software cohabiting with a worker within a pod, running a specific algorithm to determine which tasks "its" worker (the one with which it shares the pod) should perform. It also manages all interactions between the worker and the databases (retrieving/saving data, creating new tasks, etc.). A polling/scheduling agent, like a worker, exists within a single partition.

Reviewer 4: no comma before "and"
Also, I would create 2 entries : one for Polling Agent and a separate one for Scheduling Agent. Once the term Scheduling Agent gets adopted, there will be no need to have mentions of the previous name.

> **Redis:** Redis is an open source (BSD licensed), in-memory data structure store used as a database, cache, message broker, and streaming engine. Redis provides data structures such as strings, hashes, lists, sets, sorted sets with range queries, bitmaps, hyperloglogs, geospatial indexes, and streams. Redis has built-in replication, Lua scripting, LRU eviction, transactions, and different levels of on-disk persistence, and provides high availability via Redis Sentinel and automatic partitioning with Redis Cluster (see Redis (documentation)[https://redis.io/docs/about/]). In ArmoniK, Redis is used as a key-value cache for tasks data (such as payloads and results).

Reviewer 2: ```suggestion
**Redis:** Redis is an open source (BSD licensed), in-memory data structure store used as a database, cache, message broker, and streaming engine. Redis provides data structures such as strings, hashes, lists, and sorted sets with range queries, bitmaps, hyperloglogs, geospatial indexes, and streams. Redis has built-in replication, Lua scripting, LRU eviction, transactions, and different levels of on-disk persistence, and provides high availability via Redis Sentinel and automatic partitioning with Redis Cluster (see Redis (documentation)[https://redis.io/docs/about/]). In ArmoniK, Redis is used as a key-value cache for task data (such as payloads and results).
```

**Session:** Logical abstraction of a period of time during which a *client* maintains communication with ArmoniK's Control Plane in order to submit tasks, retrieve results or observe cluster behavior.

> **Submitter:** Containerized software in charge of submitting tasks, i.e. writing the corresponding data to the various databases (queue, Redis and MongoDB).

Reviewer 2: ```suggestion
**Submitter:** Containerized software in charge of submitting tasks, i.e., writing the corresponding data to the various databases (queue, Redis, and MongoDB).
```

> **Task:** Indivisible unit of computation taking one or several input data and outputting one or several results.

Reviewer: Task is launched by the client, and processed by the worker.

> **Task:** Indivisible unit of computation taking one or several input data and outputting one or several results.

Reviewer 3: Atomic! That's the term you are looking for. Moreover, tasks cannot and should not communicate between themselves.

# Glossary

## Client

User-developed software that communicates with the ArmoniK Control Plane to submit a list of tasks to be executed (by one or several workers) and retrieves results and error.

## Compute Plane

Agreed term designating the set of compute pods, i.e., the pods running a Scheduling Agent + Worker pair within the Kubernetes cluster.

## Container

A container is a standard unit of software that packages up code and all its dependencies so the application runs quickly and reliably from one computing environment to another. (see Docker [documentation](https://www.docker.com/resources/what-container/))

## Control Plane

Agreed term designating two different things depending on the context. Kubernetes and ArmoniK each have a Control Plane.
* The Kubernetes Control Plane is a set of components that manage global decision-making for the cluster, as well as the detection and management of cluster events.

> * the ArmoniK Control Plane is a set of applications that act as a bridge between the *client* and the Compute Plane. In particular, it handles communications between certain components such as the queue, Redis and MongoDB.

Reviewer: This could be split into 2 sections:

```md
## Control Plane (ArmoniK)

...

## Control Plane (Kubernetes)

...
```

## Node

In the context of Kubernetes, a node is a machine that runs containerized workloads as part of a Kubernetes cluster. A node can be a physical or virtual machine hosted in the cloud or on-premises.

## Partition

Logical segmentation of the Kubernetes cluster's pool of machines to distribute workloads according to usage. This feature is provided and handled by ArmoniK.

## Payload

Input data for a task that does not depend on any other task.

## Pod

Pods are the smallest deployable units of computing that one can create and manage in Kubernetes. A Pod is a group of one or more containers, with shared storage and network resources and a specification for how to run the containers. A Pod's contents are always co-located and co-scheduled and run in a shared context. A Pod models an application-specific "logical host": it contains one or more application containers which are relatively tightly coupled (see Kubernetes [documentation](https://kubernetes.io/docs/concepts/workloads/pods/)).

## Polling agent

Former term for scheduling agent.

## Redis

> Redis is an open source (BSD licensed), in-memory data structure store used as a database, cache, message broker and streaming engine. Redis provides data structures such as strings, hashes, lists, sets and sorted sets with range queries, bitmaps, hyperloglogs, geospatial indexes and streams. Redis has built-in replication, Lua scripting, LRU eviction, transactions and different levels of on-disk persistence and provides high availability via Redis Sentinel and automatic partitioning with Redis Cluster (see Redis [documentation](https://redis.io/docs/about/)). In ArmoniK, Redis is used as a key-value cache for task data (such as payloads and results).

Reviewer 5: ```suggestion
Redis is an open source (BSD licensed), in-memory data structure store used as a database, cache, message broker and streaming engine. Redis provides data structures such as strings, hashes, lists, sets and sorted sets with range queries, bitmaps, hyperloglogs, geospatial indexes and streams. Redis has built-in replication, Lua scripting, LRU eviction, transactions and different levels of on-disk persistence and provides high availability via Redis Sentinel and automatic partitioning with Redis Cluster (see [Redis documentation](https://redis.io/docs/about/)). In ArmoniK, Redis is used as a key-value cache for task data (such as payloads and results).
```

## Redis

Redis is an open source (BSD licensed), in-memory data structure store used as a database, cache, message broker and streaming engine. Redis provides data structures such as strings, hashes, lists, sets and sorted sets with range queries, bitmaps, hyperloglogs, geospatial indexes and streams. Redis has built-in replication, Lua scripting, LRU eviction, transactions and different levels of on-disk persistence and provides high availability via Redis Sentinel and automatic partitioning with Redis Cluster (see Redis [documentation](https://redis.io/docs/about/)). In ArmoniK, Redis is used as a key-value cache for task data (such as payloads and results).

## Scheduling agent

Containerized software cohabiting with a worker within a pod, running a specific algorithm to determine which tasks "its" worker (the one with which it shares the pod) should perform. It also manages all interactions between the worker and the databases (retrieving/saving data, creating new tasks, etc.), as well as managing worker errors and retrying/resubmitting failed tasks when necessary. A scheduling agent, like a worker, exists within a single partition.

## Session

> A session is a logical container for tasks and associated data (task statut, results, errors, etc). Every task is submitted within a session. An existing session can be resumed to retrieve data or submit new tasks. When a session is cancelled, all associated executions still in progress are interrupted.

Reviewer 4: Is it really a container? I would start with " A session is a limited time when two systems communicate with each other. "

Author: In fact a session doesn't refer to a limited amount of time. So far session are eternal and never deleted. One can still retrieve data from an old session and submit new tasks.
